### PR TITLE
prevent drop of default index values

### LINKED ISSUE #422: Merge on non default index drops index values
```
2.3.3 :127 > df1
 => #<Daru::DataFrame(10x3)>
                 color director_n num_critic
    Avatar       Color James Came        723
 Pirates of      Color Gore Verbi        302
   Spectre       Color Sam Mendes        602
 The Dark K      Color Christophe        813
 Star Wars:        nil Doug Walke        nil
 John Carte      Color Andrew Sta        462
 Spider-Man      Color  Sam Raimi        392
   Tangled       Color Nathan Gre        324
 Avengers:       Color Joss Whedo        635
 Harry Pott      Color David Yate        375 
2.3.3 :128 > df2
 => #<Daru::DataFrame(10x3)>
            imdb_score aspect_rat movie_face
    Avatar         7.9       1.78      33000
 Pirates of        7.1       2.35          0
   Spectre         6.8       2.35      85000
 The Dark K        8.5       2.35     164000
 Star Wars:        7.1        nil          0
 John Carte        6.6       2.35      24000
 Spider-Man        6.2       2.35          0
   Tangled         7.8       1.85      29000
 Avengers:         7.5       2.35     118000
 Harry Pott        7.5       2.35      10000 
2.3.3 :129 > df1.merge(df2)
 => #<Daru::DataFrame(10x6)>
                 color director_n num_critic imdb_score aspect_rat movie_face
          0      Color James Came        723        7.9       1.78      33000
          1      Color Gore Verbi        302        7.1       2.35          0
          2      Color Sam Mendes        602        6.8       2.35      85000
          3      Color Christophe        813        8.5       2.35     164000
          4        nil Doug Walke        nil        7.1        nil          0
          5      Color Andrew Sta        462        6.6       2.35      24000
          6      Color  Sam Raimi        392        6.2       2.35          0
          7      Color Nathan Gre        324        7.8       1.85      29000
          8      Color Joss Whedo        635        7.5       2.35     118000
          9      Color David Yate        375        7.5       2.35      10000
```

### PR DESCRIPTION
**Pull request by rohitner**
made some minor changes for fixing #422

### CODE REVIEW

**Comment by rohitner** [comment]
it checks if the dataframes have same indices and sets the same index in the merged dataframe, which was earlier 0 1 2 3 ..

**Comment by zverok** [comment]
Yes, I've almost guessed :) But any bugfix should first have a showcase of problem fixed in specs. It is important, because any behavior not shown in specs can be easily broken again on the next iteration.

**Comment by rohitner** [comment]
@zverok is this the right way?

File: spec/dataframe_spec.rb [hunk 1]
```diff
@@ -3243,6 +3243,15 @@
       it { expect(subject['a'].to_a).to eq [1, 2, 3] }
       it { expect(subject[:b].to_a).to eq [4, 5, 6] }
     end
+
+    context "preserves indices for dataframes with same index" do
+      index = ['one','two','three']
```
Comment (zverok) [inline spec/dataframe_spec.rb]: Please use `let` for index.

File: spec/dataframe_spec.rb [hunk 2]
```diff
@@ -3243,6 +3243,15 @@
...
+      let(:df1) { Daru::DataFrame.new({ 'a' => [1, 2, 3], 'b' => [3, 4, 5] }, index: index) }
+      let(:df2) { Daru::DataFrame.new({ 'c' => [4, 5, 6], 'd' => [7, 8, 9] }, index: index) }
+      subject { df1.merge df2 }
+
+      it { expect(subject.index.to_a).to eq index }
```
Comment (zverok) [inline spec/dataframe_spec.rb]:
```ruby
its(:index) { is_expected.to eq Daru::Index.new(index) }
```
...ro something like that.

**Comment by rohitner** [comment]
done! learning a lot these days by contributing :smile:

File: lib/daru/dataframe.rb [hunk 3]
```diff
@@ -1824,14 +1824,15 @@ def merge other_df # rubocop:disable Metrics/AbcSize
...
-
-      DataFrame.new({}, order: new_fields).tap do |df_new|
+      temp = DataFrame.new({}, order: new_fields).tap do |df_new|
         (0...nrows).each do |i|
           df_new.add_row row[i].to_a + other_df.row[i].to_a
         end
 
         df_new.update
       end
+      temp.index= @index.to_a if @index.to_a == other_df.index.to_a
```
Comment (athityakumar) [inline lib/daru/dataframe.rb]: Aren't the `#to_a`s required? Wouldn't this work?

```ruby
temp.index = @index if @index == other_df.index
```

Comment (rohitner) [reply lib/daru/dataframe.rb]: yaah its does :sweat_smile:

File: lib/daru/dataframe.rb [hunk 4]
```diff
@@ -1824,14 +1824,15 @@ def merge other_df # rubocop:disable Metrics/AbcSize
...
-      DataFrame.new({}, order: new_fields).tap do |df_new|
+      temp = DataFrame.new({}, order: new_fields).tap do |df_new|
         (0...nrows).each do |i|
           df_new.add_row row[i].to_a + other_df.row[i].to_a
         end
 
         df_new.update
       end
+      temp.index= @index if @index == other_df.index
+      temp
```
Comment (zverok) [inline lib/daru/dataframe.rb]: Oh, sorry for not noticing it earlier. You can reuse that `tap` block that was here before (it is exactly the idiom for avoiding "temp" local variables:
```ruby
DataFrame.new({}, order: new_fields).tap do |df_new|
  (0...nrows).each do |i|
    df_new.add_row row[i].to_a + other_df.row[i].to_a
  end

  df_nex.index = @index if @index == other_df.index

  df_new.update
end
```

Comment (rohitner) [reply lib/daru/dataframe.rb]: oh got it! :+1:

**Pull request merged**